In [22]:
pip install faiss-cpu -q

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [1]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [ ]:
!pip install transformers==4.51.3

In [8]:
pip install sentence-transformers

Defaulting to user installation because normal site-packages is not writeable
  Using cached sentence_transformers-5.0.0-py3-none-any.whl (470 kB)
  Using cached huggingface_hub-0.33.4-py3-none-any.whl (515 kB)
  Using cached tqdm-4.67.1-py3-none-any.whl (78 kB)
  Using cached pillow-11.3.0-cp39-cp39-macosx_11_0_arm64.whl (4.7 MB)
  Using cached scipy-1.13.1-cp39-cp39-macosx_12_0_arm64.whl (30.3 MB)
  Using cached transformers-4.53.2-py3-none-any.whl (10.8 MB)
  Using cached torch-2.7.1-cp39-none-macosx_11_0_arm64.whl (68.6 MB)
  Using cached scikit_learn-1.6.1-cp39-cp39-macosx_12_0_arm64.whl (11.1 MB)
  Using cached hf_xet-1.1.5-cp37-abi3-macosx_11_0_arm64.whl (2.6 MB)
  Using cached fsspec-2025.7.0-py3-none-any.whl (199 kB)
  Using cached PyYAML-6.0.2-cp39-cp39-macosx_11_0_arm64.whl (172 kB)
  Using cached requests-2.32.4-py3-none-any.whl (64 kB)
  Using cached filelock-3.18.0-py3-none-any.whl (16 kB)
     |████████████████████████████████| 1.6 MB 1.4 MB/s eta 0:00:01
     |█████████

In [3]:
! pip install streamlit==1.35.0
! pip install pandas==2.2.2
! pip install sentence-transformers==2.6.1
! pip install faiss-cpu==1.7.4
! pip install numpy==1.24.4
! pip install openai==0.28.0


  Using cached faiss-cpu-1.7.4.tar.gz (57 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  error: subprocess-exited-with-error
  
  × Building wheel for faiss-cpu (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [21 lines of output]
      /private/var/folders/vw/z88nn85s59nbjgrz304n0h7h0000gp/T/pip-build-env-gruvjv_v/overlay/lib/python3.12/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!
      
              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:
      
              License :: OSI Approved :: MIT License
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ****************************************

In [51]:
pip install --upgrade openai


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
  Using cached openai-1.97.0-py3-none-any.whl (764 kB)
  Attempting uninstall: openai
    Found existing installation: openai 0.28.0
    Uninstalling openai-0.28.0:
      Successfully uninstalled openai-0.28.0
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [41]:
pip install openai==0.28

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 76 kB 1.8 MB/s eta 0:00:01
     |████████████████████████████████| 467 kB 1.6 MB/s eta 0:00:01
     |████████████████████████████████| 44 kB 2.7 MB/s eta 0:00:01
     |████████████████████████████████| 43 kB 2.5 MB/s eta 0:00:01
     |████████████████████████████████| 89 kB 1.8 MB/s eta 0:00:01
     |████████████████████████████████| 63 kB 1.7 MB/s eta 0:00:01
     |████████████████████████████████| 47 kB 2.6 MB/s eta 0:00:01
  Attempting uninstall: openai
    Found existing installation: openai 1.97.0
    Uninstalling openai-1.97.0:
      Successfully uninstalled openai-1.97.0
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


# Large Language Model & Retrieval Augmented Generation

- Apa itu LLM (Large Language Model)?

LLM adalah model kecerdasan buatan yang dilatih dengan banyak teks untuk bisa mengerti, menulis, dan menjawab pertanyaan dalam bahasa manusia. LLM ini bisa menjawab pertanyaan, bantu menulis, merangkum teks, dan banyak hal lainnya.

- Apa itu RAG (Retrieval-Augmented Generation)?

RAG adalah cara untuk menggabungkan kekuatan LLM + data kamu sendiri. Atau sederhananya  **RAG = cari data (Retrieval) + jawab pakai LLM (Generation)**

- Kapan butuh RAG?
    - Kalau LLM tidak tahu data kamu (misalnya data internal perusahaan)
    - Kalau kamu ingin jawaban berbasis bukti/data, bukan cuma jawaban umum

## Alur RAG

Dalam melakukan RAG, terdapat beberapa tahapan yang perlu dilakukan dari data tabular yang dimiliki agar nantinya model LLM dapat memberikan jawaban yang sesuai. Berikut beberapa tahapannya. 

### Step 1: Mempersiapkan Data Tabular

Agar nantinya LLM dapat menghasilkan jawaban yang relevan berdasarkan data, kita perlu mempersiapkan data dalam bentukan tabular terlebih dahulu, dimana bentukan data tabular yang dimaksud di sini adalah data yang berisikan kolom-kolom data dan juga baris.

Bentukan format data yang digunakan bisa bervariasi seperti excel, csv, atau lainnya.

In [2]:
import pandas as pd
product = pd.read_csv("/Users/wrahma01/Documents/Building LLM Applications for Structured Data Insights/data/products.csv")
product

id   nama_produk    kategori     harga  stok
0   1      Laptop X  Elektronik  12000000    15
1   2   Headphone Y  Elektronik   1500000     5
2   3  Smartphone Z  Elektronik   8000000     3
3   4     Monitor A  Elektronik   3000000     8
4   5    Keyboard B   Aksesoris    500000    25

### Step 2: Transformasi Data ke Format Teks

Bentukan data tabular merupakan bentukan yang suliit dipelajari oleh LLM dikarenakan setiap kolomnya berdiri secara individual, agar nantinya lebih mudah untuk dipelajari, tahapan yang perlu dilakukan adalah mengubah 1 baris data yang terpisah antara setiap kolom, menjadi 1 kalimat penuh.

Berikut adalah contohnya:

**Data Tabular**
| Nama | Gaji | Divisi  |
| ---- | ---- | ------- |
| Rina | 10jt | Finance |

**Hasil Transformasi**
```python
Nama: Rina | Gaji: 10jt | Divisi: Finance
```

Dengan bentuk ini, LLM bisa "membaca" dan "menalar" seperti membaca kalimat biasa. Selain dari itu, LLM jadi tahu bahwa “Rina” berhubungan dengan “Finance” dan “10jt” — hal yang tidak otomatis dipahami kalau datanya hanya angka-angka di tabel.

Terdapat sebuah fungsi `.agg()` yang biasanya digunakan untuk melakukan perhitungan, fungsi tersebut dapat kita manfaatkan lebih lanjut untuk menggabungkan informasi dari beberapa kolom. 

Fungsi tersebut nantinya akan digabungkan dengan metode `.join()`, metode `.join()` digunakan untuk menggabungkan string, bukan untuk menggabungkan DataFrame seperti merge(). 

Agar nantinya semua fungsi tersebut bisa digabungkan, satu syarat yang harus dipenuhi adalah semua kolom harus memiliki tipe data yang sama. Hal tersebut bisa dipenuhi dengan menggunakan fungsi `.astype()`.

```python
df[['kolom_1', 'kolom_2']].astype('tipe_data').agg(' | '.join, axis = 1)
```

In [4]:
product.head(1)

id nama_produk    kategori     harga  stok
0   1    Laptop X  Elektronik  12000000    15

In [5]:
# Mari gabungkan 2 buah kolom terlebih dahulu 
product[['nama_produk', 'harga']].astype('str').agg(' | '.join, axis = 1)

0       Laptop X | 12000000
1     Headphone Y | 1500000
2    Smartphone Z | 8000000
3       Monitor A | 3000000
4       Keyboard B | 500000
dtype: object

Setelah berhasil memahami bagaimana cara kerja menggabungkan beberapa informasi data menjadi satu kesatuan, tahapan berikutnya adalah kita harus mempersiapkan fungsi untuk kedepannya bisa menerima semua inputan data ataupun kolom yang ingin diproses.

Hal ini diperlukan karena tidak tentu semua kolom harus diproses atau diambil untuk dipahami.

Berikut kerangka berfikir dalam pembuatan fungsinya:
- Siapan sebuah fungsi yang bisa menampung 2 parameter
    + Parameter untuk menampung variabel dataframe
    + Parameter untuk memilih kolom-kolom yang diperlukan
- Memanfaatkan fungsi yang sudah dibuat tadi, tapi ada hal yang perlu diganti, yaitu menyesuapkan nama variabel dataframe dan nama kolom, sesuai dengan parameter yang dipersiapkan


In [6]:
# membuat sebuah fungsi
def penggabungan_kolom(df, kolom_data):
    df['teks'] = df[kolom_data].astype('str').agg(' | '.join, axis = 1)
    return df

In [7]:
penggabungan_kolom(df = product, kolom_data=['id', 'nama_produk', 'kategori','harga', 'stok'])

id   nama_produk    kategori     harga  stok  \
0   1      Laptop X  Elektronik  12000000    15   
1   2   Headphone Y  Elektronik   1500000     5   
2   3  Smartphone Z  Elektronik   8000000     3   
3   4     Monitor A  Elektronik   3000000     8   
4   5    Keyboard B   Aksesoris    500000    25   

                                          teks  
0    1 | Laptop X | Elektronik | 12000000 | 15  
1   2 | Headphone Y | Elektronik | 1500000 | 5  
2  3 | Smartphone Z | Elektronik | 8000000 | 3  
3     4 | Monitor A | Elektronik | 3000000 | 8  
4     5 | Keyboard B | Aksesoris | 500000 | 25

### Step 3: Mencari Kesamaan Teks

####  Step 3.1: Embeddng Teks

Pada tahapan yang dilakukan adalah mengubah teks menjadi representasi angka (vektor) yang menangkap makna atau arti dari teks tersebut. Makna atau arti yang dimaksud di sini adalah untuk mengetahui kedekatan atau kemiripan antara satu data dengan data lainnya.

Informasi mengenai kemiripan antara satu data dengan yang lain penting dikarenakan tujuan awal dari RAG adalah mengambil informasi data yang relevan dengan pertanyaan yang diberikan, untuk menghasilkan jawaban LLM yang relevan. 

**Model: paraphrase-MiniLM-L6-v2**, model ini adalah bagian dari library Sentence Transformers, yang dibangun di atas BERT dan sangat cocok untuk melakukan embedding teks.

Dari bantuan di atas, hal yang dapat dilakukan adalah setiap kalimat akan diubah menjadi bentukan vektor. Bentukan vektor tersebut akan dihasilkan akan tergantung dengan semua data train yang dimiliki oleh model **paraphrase-MiniLM-L6-v2**.

In [9]:
# mempersiapkan model embedding yang akan digunakan
from sentence_transformers import SentenceTransformer, util

model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

/Users/wrahma01/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(
/Users/wrahma01/Library/Python/3.9/lib/python/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [10]:
# contoh kata output yang akan digunakan
corpus_query = "How I learn AI?"

# contoh kata output yang akan digunakan
corpus_ouput = [
    "Saya cinta machine learning.",
    "Python adalah bahasa yang keren.",
    "Deep learning meningkatkan klasifikasi gambar.",
    "Artificial intelligence adalah masa depan.",
    "Daftar di Akademi AI!"
]

Agar nantinya bisa diubah menjadi bentukan vector, fungsi yang akan digunakan adalah `encode()` dan terdapat parameter yang harus diisi `convert_to_numpy=True` tujuannya untuk output dari embedding akan dikembalikan dalam bentuk numpy.ndarray.

In [11]:
# mempersiapkan model embedding yang akan digunakan
from sentence_transformers import SentenceTransformer, util

model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

In [12]:
embedding_query = model.encode(corpus_query, convert_to_numpy= True)
embedding_query

array([ 1.30131885e-01, -4.97135371e-01, -1.91070244e-01, -8.16830933e-01,
        4.97040808e-01,  5.74899077e-01,  1.03566611e+00, -2.33822986e-01,
       -5.72757013e-02,  1.06510438e-01,  3.91608775e-01,  1.06586829e-01,
        1.69848278e-01,  1.16042174e-01, -2.48942941e-01, -2.52282709e-01,
       -3.64941627e-01,  4.29967076e-01, -2.83076078e-01, -3.97258341e-01,
       -2.01484099e-01, -2.92198211e-01,  6.35252148e-02,  1.02740742e-01,
        4.27272797e-01,  7.47818291e-01,  1.51754931e-01, -1.67966023e-01,
        5.46775222e-01, -4.32285458e-01,  8.72432411e-01, -4.22608107e-01,
        6.00874364e-01, -1.04316071e-01, -1.77313581e-01, -4.81827199e-01,
       -5.37118077e-01, -1.13151617e-01, -2.07263306e-01, -1.95097759e-01,
       -4.01369750e-01,  3.54159087e-01,  2.07943171e-01, -2.64321864e-01,
        8.45733762e-01, -1.09710440e-03,  1.06256127e-01, -3.91945213e-01,
        5.07670403e-01,  2.36963525e-01, -5.72868645e-01, -1.15839608e-01,
        1.36171564e-01,  

In [13]:
embedding_output = model.encode(corpus_ouput, convert_to_numpy= True)

Dari semua kalimat yang berada di atas, nantinya akan coba dicari kesamaanya dengan menghitung *cosine similarity*. 

- **Semakin mendekati 1**, maka akan dianggap semakin mirip
- **Semakin mendekati 0**, maka akan dianggap semakin tidak mirip

Fungsi yang digunakan adalah `util.cos_sim()`

In [14]:
cosine_scores = util.cos_sim(embedding_query, embedding_output)
cosine_scores

tensor([[0.3704, 0.0312, 0.2236, 0.4614, 0.4938]])

In [15]:
corpus_query

'How I learn AI?'

In [16]:
corpus_ouput

['Saya cinta machine learning.',
 'Python adalah bahasa yang keren.',
 'Deep learning meningkatkan klasifikasi gambar.',
 'Artificial intelligence adalah masa depan.',
 'Daftar di Akademi AI!']

Metode di atas dinamakan **Pencarian Semantik**, karena tujuan dari pencariannya adalah melihat kontekstual satu kalimat, tidak hanya berdasarkan apakah sebuah kata itu ada atau tidak.

> Bagaimana nantinya kalo kita coba terapkan pada data kita? Data manakan yang paling mendekati, jika diberikan pertanyaan **Laptop untuk bekerja?**

In [17]:
# persiapan query & embedding baru
query =  "laptop untuk bekerja"
embedding_query_baru = model.encode(query, convert_to_numpy=True)

In [18]:
# persiapan embedding untuk data lama
embedding_dataframe = model.encode(product['teks'], convert_to_numpy=True)

In [19]:
# menghtiung cosine similarity
cosine_scores = util.cos_sim(embedding_query_baru, embedding_dataframe)
cosine_scores

tensor([[0.4676, 0.3058, 0.2649, 0.3067, 0.2315]])

In [20]:
product['teks']

0      1 | Laptop X | Elektronik | 12000000 | 15
1     2 | Headphone Y | Elektronik | 1500000 | 5
2    3 | Smartphone Z | Elektronik | 8000000 | 3
3       4 | Monitor A | Elektronik | 3000000 | 8
4       5 | Keyboard B | Aksesoris | 500000 | 25
Name: teks, dtype: object

### Step 3.2: FAISS Indexing Dari Pertanyaan 

FAISS (Facebook AI Similarity Search) adalah library dari Facebook AI Research yang dirancang untuk pencarian cepat terhadap vektor berdimensi tinggi. FAISS sangat sering digunakan dalam aplikasi seperti semantic search, image similarity, dan retrieval dalam RAG (Retrieval-Augmented Generation).

**Cara Kerja FAISS**

1. Mengambil nilai vektor dari hasil embedding
2. Melakukan pembuatan index & menambahkan informasi index
3. Melakukan pencarian berdasarkan pertanyaan yang diberikan

Dikarenakan untuk melakukan indexing yang menggabungkan antara cosine similarity dan metode FAIIS tidak bisa secara langsung, melainkan harus dilakukan secara manual. 

Maka dari itu mari kita coba buat perhitungan cosine similarity secara manual.

In [ ]:
import faiss
import numpy as np

# melakukan perhitungan
embedding_dataframe = embedding_dataframe / np.linalg.norm(embedding_dataframe, axis=1, keepdims=True)
embedding_dataframe = embedding_dataframe.astype('float32')

# mengambil nilai embedding
dimension = embedding_dataframe.shape[1]
dimension

384

Untuk penambahan informasi index, bisa menggunakan fungsi `IndexFlatL2()` dan hasilnya akan ditambahkan ke hasil embedding dengan menggunakan `.add()`

In [25]:
index = faiss.IndexFlatL2(dimension)
index.add(embedding_dataframe)

Dari hasil dimensinya jika dibandingkan kembali dengan contoh query pertanyaan di atas, akan seperti di bawah ini.

In [26]:
query

'laptop untuk bekerja'

In [27]:
embedding_index_query = model.encode([query])

In [28]:
D, I = index.search(embedding_index_query, k = 2)

In [29]:
D

array([[44.40567 , 46.680927]], dtype=float32)

In [30]:
I

array([[0, 3]])

In [31]:
cosine_scores

tensor([[0.4676, 0.3058, 0.2649, 0.3067, 0.2315]])

- `D:` jarak (distance) antara query_vector dengan vektor-vektor terdekat.
- `I:` indeks dari vektor-vektor yang paling mirip — ini yang kamu cari.

Mari kita gabungkan perhitungan embeddings sampai dengan proses pembuatan index menjadi sebuaah fungsi.

Berikut kerangka berfikir dalam pembuatan fungsinya:

- Siapan sebuah fungsi yang bisa menampung teks dari data
- Melakukan perhitungan cosine similarity
- Melakukan indexing
     - Mengambil informasi shape
     - Melakukan pembutan index berdasarkan hasil embedding cosine similarity
     - Menerapkan hasil index ke embediings

In [32]:
# membuat sebuah fungsi
def build_faiss_index_cosine(teks):
    # Bagian untuk melakukan embeddings
    embedding = model.encode(teks , convert_to_numpy=True)

    # Melakukan perhitungan cosine
    embedding = embedding / np.linalg.norm(embedding, axis=1, keepdims=True)
    embedding = embedding.astype('float32')

    # Melakukan indexing
    dimension = embedding.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embedding)

    return index, embedding

[Optional] Pengambilan sumber informasi berdasarkan berapa banyak informasi yang paling dekat

In [33]:
# membuat sebuah fungsi
def retrieve(query, index, df, top_k=3):
    # 1. Encode dan normalisasi query
    query_embedding = model.encode([query], convert_to_numpy=True)
    query_embedding = query_embedding / np.linalg.norm(query_embedding, axis=1, keepdims=True)
    query_embedding = query_embedding.astype("float32")

    # 2. Search ke FAISS
    scores, indices = index.search(query_embedding, top_k)

    # 3. Ambil baris data sesuai hasil indexing
    result_df = df.iloc[indices[0]].copy()
    result_df['similarity_score'] = scores[0]

    return result_df

# Bagaimana LLM Bekerja [Optional]

## 1. Bagaimana Model "Belajar" dari Pertanyaan?

Model LLM tidak belajar dari data setelah dilatih. Tapi ia memahami dan menafsirkan pertanyaan berdasarkan *pretraining* yang sudah dilakukan atas triliunan token.

```
Bagaimana cara belajar Python untuk pemula?
```

Model akan:

- Memecah pertanyaan menjadi token
    + Token adalah unit terkecil dari teks yang digunakan oleh model bahasa seperti GPT saat membaca dan memproses input. Token bisa berupa:
        * Kata penuh → seperti "Python"
        * Sebagian kata → seperti "Py", "thon"
        * Tanda baca → seperti ",", ".", atau "!"
        * Spasi juga bisa dihitung sebagai token
- Menggunakan representasi internal yang sangat besar (disebut embedding) untuk memahami struktur kalimat, maksud, dan konteks

Contohnya, kata "belajar Python" akan dikenali sebagai sesuatu yang mengarah ke programming, tutorial, atau konsep dasar coding.

## 2. Mencocokkan ke Data yang Paling Dekat

```
[User Query] → "Apa manfaat dari vitamin C?"
         ↓
[Query Embedding]
         ↓
[Search via FAISS / Vector DB]
         ↓
[Temukan dokumen yang mirip secara makna]
         ↓
[Gabungkan query + dokumen → kirim ke GPT-4.1-mini]
         ↓
[Jawaban: "Vitamin C penting untuk sistem imun dan antioksidan."]
```

## 3. Bagaimana LLM Menjawab?

Setelah dokumen hasil pencarian diberikan (biasanya sebagai context), model akan:

- Memproses seluruh input (pertanyaan + dokumen)
- Menggunakan kemampuannya untuk:
    - Menemukan bagian yang paling relevan
    - Menyusun kalimat berdasarkan jawaban yang konsisten secara logika dan bahasa
- Tidak mengambil langsung dari dokumen, tapi merangkai ulang dengan pemahaman model

### 3.1 Proses Merangkai Kalimat

Proses merangkai kata-kata jawaban akan berdasarkan perhitungan jarak dan hasil probabilitas yang paling sesuai dari hasil data-data latih yang model miliki, ditambahkan dengan data-data baru yang diinput.

- Untuk perhitungan jarak, konsep yang akan digunakan akan sama dengan konsep indexing dengan metode FAISS
- Perhitungan probabilitas akan memanfaatkan metode klasifikasi dalam menentukan kata-kata yang paling mungkin untuk dimanfaatkan untuk melanjutkan jawaban.

### 3.2 Kapan Akan Berhenti Merangkai Jawaban

Model LLM akan berhenti dalam memberikan jawaban ketika, model sudah mencapai batas token dalam merangkai sebuah kalimat. Sebagai contoh,

- GPT:
    - Model tertua GPT‑3 hanya 2.048 token
    - GPT‑3.5‑Turbo umum: 4K, versi “16k” dukung hingga 16K
    - GPT‑4 basic: 8K; varian 32k: 32K
    - GPT‑4 Turbo: 128K input, output tetap 4K
    - GPT‑4.1: hingga 1 juta token total

- LLaMA (Meta)
    - LLaMA 1 & 2: sekitar 2 048–4 096 token 
    - LLaMA 3 (8B, 70B, 405B): hingga 128 000 token 
    - LLaMA 3.2/3.3: tetap 128 000 token 
    - LLaMA 4 Scout: 10 juta token 
    - LLaMA 4 Maverick: 1 juta token

- Gemini:
    - Gemini 1 (Nano, dst.): 32 768 token 
    - Gemini 1.5 Flash/Pro:
        - Flash standar: 128 000 token
        - Pro (private/enterprise): hingga 1 000 000 token; kini sudah mulai buka akses hingga 2 000 000 token 
    - Gemini 2.5 Flash: 1 048 576 input, output hingga 65 536 token

- DeepSeek
    - DeepSeek‑LLM (7B & 67B): 4 096 token 
    - DeepSeek‑Coder: 16 384 token
    - DeepSeek‑R1/V2/V3:
        - R1/V2: 128 000 token 
        - V3: konteks 128 000 token

Selain dari batasan token, model LLM akan selalu memprediksi sebuah probabilitas yang diberikan `End Of Sequence`/`<eos>`. Ketika hasil probabilitas untuk kata berikutnya adalah `<eos>`, maka model akan memberhentikan dalam membuat jawaban, walaupun tokennya masih tersedia banyak. 

Mengapa EOS Penting?
- fisiensi: Model tidak terus memproses output yang tidak dibutuhkan.
- ualitas teks: Mencegah model "mengoceh" atau menghasilkan teks yang tidak relevan.
- ontrol output: Digunakan saat menyusun pipeline NLP seperti summarization, translation, dll.

In [34]:
# query: sebuah tempat untuk kita nantinya memberikan sebuah inputan/pertanyaan
# context: untuk meberikan data yang ingin coba dipelajari atau diproses 
# api_key: untuk kita memasukan key, untuk mengakses model generative ai secara online
def generate_answer(query, context, api_key):
    # untuk memasukan api key dari generative model yang digunakan
    openai.api_key = api_key
    # untuk memberitahu secara spesifik apa yang perlu dilakukan oleh model generative ai
    system_message = "Kamu adalah asisten cerdas yang menjawab pertanyaan berdasarkan data yang diberikan."
    # untuk user memberikan inputan pertanyaan ataupun data yang ingin dipelajari
    user_message = f"""
    Pertanyaan: {query}

    Data yang relevan:
    {context}
    """
    response = openai.ChatCompletion.create(
        model = 'gpt-4.1-mini', # model yang digunakan
        # sysmtem messages atau untuk mengolah inputan data ataupun user
        messages = [
            {"role": "system", "content": system_message},
            {'role': "user", "content": user_message}
        ],
        # untuk mengatur tingkat pemilihan prediksi kata berikutnya
        temperature = 0.3,
        # untuk mengatur jumlah token yang bisa diproses
        max_tokens = 1000
    )

In [42]:
import openai

Temperature 

- 0 -> Model generative memilih probabilitas yang paling tinggi 
    + aku -> 88%
    + saya -> 89% -> yang dipilih ini
- 0.5 -> Model generative memilih probabilitas bukan yang paling tinggi, top 3 - top 5
    + aku -> 88%
    + saya -> 89%
    + gue -> 74% -> yang dipilih ini
    + abdi - 70%
- 1.2 -> model generative memilih probabilitas semakin random
    * Pilihan kata 1
        + aku -> 88%
        + saya -> 89% -> ini yang dipilih
        + gue -> 74% 
        + abdi - 70%
    * Pilihan kata 2
        + adalah 99%
        + apel 40% -> ini yang dipilih
        + manusia 98%
    

In [54]:
import openai

def generate_answer(query, context, api_key):
    openai.api_key = api_key
    system_message = "Kamu adalah asisten cerdas yang menjawab pertanyaan berdasarkan data yang diberikan."
    user_message = f"""
    Pertanyaan: {query}

    Data yang relevan:
    {context}
    """
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": system_message},
            {"role": "user", "content": user_message}
        ],
        temperature=0.3,
        max_tokens=1000
    )
    return response.choices[0].message["content"]

In [ ]:
import openai

def generate_answer(query, context, api_key):
    # Untuk versi OpenAI API terbaru (1.x.x ke atas), inisialisasi client seperti ini
    client = openai.OpenAI(api_key=api_key)

    system_message = "Kamu adalah asisten cerdas yang menjawab pertanyaan berdasarkan data yang diberikan."
    user_message = f"""
    Pertanyaan: {query}

    Data yang relevan:
    {context}
    """
    response = client.chat.completions.create(  # Perubahan di sini
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": system_message},
            {"role": "user", "content": user_message}
        ],
        temperature=0.3,
        max_tokens=1000
    )
    return response.choices[0].message.content # Perubahan di sini untuk mengakses konten

# Contoh penggunaan:
# Pastikan Anda mengganti 'YOUR_API_KEY' dengan kunci API OpenAI Anda yang sebenarnya.
# Saya telah mengosongkan api_key di sini karena ini adalah contoh publik.
# Anda perlu mengisi api_key Anda saat menjalankan kode ini.
print(generate_answer(query= 'manakah gunung tertinggi di dunia?',
                context= f""" There are at least 108 mountains on Earth with elevations of 7,200 m (23,622 ft; 4 mi) or greater above sea level. Of these, 14 are more than 8,000 m (26,247 ft; 5 mi).[1] The vast majority of these mountains are part of either the Himalayas or the Karakoram mountain ranges located on the edge of the Indian Plate and Eurasian Plate in China, India, Nepal, and Pakistan.""",
                api_key= ''))

Gunung tertinggi di dunia adalah Gunung Everest, dengan ketinggian sekitar 8.848 meter (29.029 kaki) di atas permukaan laut. Gunung Everest terletak di Pegunungan Himalaya, di perbatasan antara Nepal dan Tibet.
